donkeycar/parts/pins.py

```diff
@@ -0,0 +1,913 @@
+
+
+#
+# ----- RPi.GPIO/Jetson.GPIO implementations -----
+#
+try:
+    import RPi.GPIO as GPIO
+    # lookups to convert abstact api to GPIO values
+    gpio_pin_edge = [None, GPIO.RISING, GPIO.FALLING, GPIO.BOTH]
+    gpio_pin_pull = [None, GPIO.PUD_OFF, GPIO.PUD_DOWN, GPIO.PUD_UP]
+    gpio_pin_scheme = {PinScheme.BOARD: GPIO.BOARD, PinScheme.BCM: GPIO.BCM}
+except ImportError:
+    print("pigpio was not imported.")
+    globals()["GPIO"] = None
+
+
+def gpio_fn(pin_scheme, fn):
+    """
+    Convenience method to call GPIO function
+    using desired pin scheme.  This restores
+    the previous pin scheme, so it is safe to 
+    mix pin schemes.
+    """
+    prev_scheme = GPIO.getmode() or pin_scheme
+    GPIO.setmode(pin_scheme)
+    val = fn()
+    GPIO.setmode(prev_scheme)
+    return val
+
+
+class InputPinGpio(InputPin):
+    def __init__(self, pin_number:int, pin_scheme:str, pull=PinPull.PULL_NONE) -> None:
+        """
+        Input pin ttl HIGH/LOW using RPi.GPIO/Jetson.GPIO
+        pin_number: GPIO.BOARD mode point number
+        pull: enable a pull up or down resistor on pin.  Default is PinPull.PULL_NONE
+        """
+        self.pin_number = pin_number
+        self.pin_scheme = gpio_pin_scheme[pin_scheme]
+        self.pull = pull
+        self.on_input = None
+        self._state = PinState.NOT_STARTED
+        super().__init__()
+
+    def _callback(self, pin_number):
+        if self.on_input is not None:
+            self.on_input(self.pin_number, self.input())
+
+    def start(self, on_input=None, edge=PinEdge.RISING) -> None:
+        """
+        on_input: function to call when an edge is detected, or None to ignore
+        edge: type of edge(s) that trigger on_input; default is 
+        """
+        if self.state() != PinState.NOT_STARTED:
+            raise RuntimeError("Attempt to start InputPin that is already started.")
+        gpio_fn(self.pin_scheme, lambda: GPIO.setup(self.pin_number, GPIO.IN, pull_up_down=gpio_pin_pull[self.pull]))
+        if on_input is not None:
+            self.on_input = on_input
+            gpio_fn(self.pin_scheme, lambda: GPIO.add_event_detect(self.pin_number, gpio_pin_edge[edge], callback=self._callback))
+        self.input()  # read first state
+
+    def stop(self) -> None:
+        if self.state() != PinState.NOT_STARTED:
+            gpio_fn(self.pin_scheme, lambda: GPIO.cleanup(self.pin_number))
+            self._state = PinState.NOT_STARTED
+
+    def state(self) -> int:
+        return self._state
+
+    def input(self) -> int:
+        self._state = gpio_fn(self.pin_scheme, lambda: GPIO.input(self.pin_number))
+        return self._state
+
+
+class OutputPinGpio(OutputPin):
+    """
+    Output pin ttl HIGH/LOW using Rpi.GPIO/Jetson.GPIO
+    """
+    def __init__(self, pin_number:int, pin_scheme:str) -> None:
+        self.pin_number = pin_number
+        self.pin_scheme = gpio_pin_scheme[pin_scheme]
+        self._state = PinState.NOT_STARTED
```

Comment:
I think all of these members are initiated equally in all of the derived classes, so maybe move this part of the construction into the base class?

Reply:
I generally do not like to use inheritance as a way to share code; I would prefer composition.  However, even that is not warranted here as these initializations are actually a little different except for a couple of cases.  Trying to share this simple code would create coupling that is probably worse than the cure.

Reply:
I went and reviewed the RPi.GPIO source code and found that the method that sets the mode will only allow it to be set once!  See https://github.com/Tieske/rpi-gpio/blob/c36c6101712d5a0c370ef0023aae237b734bf51f/source/py_gpio.c#L514  So that means we should not actually be using mixed modes.  It's also been bugging me a little that using different libraries; PiGPIO and RPi.GPIO can get you in trouble.  So we may want to have a configuration that chooses the library and the pin scheme.  We would initialize the pin provider api with that configuraiton.  Pin specs will still include the scheme, but it much match the configuration or a runtime exception would be raised.  In this way we would prevent mixed libraries and mixed pin schemes.  Further, we could get rid of the code that tries to get/set/reset the pin mode, which causes a race condition.  What do you think @DocGarbanzo?

Reply:
The other thing I could do, which would protect if other code is also using RPi.GPIO, is to check the mode to see if it is already set to a different mode and raise and error if it is.  This is essentially what the library does, but we can do it in python and give a better error message.  This will also detect if other code is using RPi.GPIO in a different mode.  I'll probably do that.  I need to make sure I go back and make sure all the documentation says all pins must be in the same mode.

Reply:
So I changes the code so that it would detect any attempt to mix pin modes and raise an error.  So, unless there was no pin mode set yet, it will never try to change the pin mode.  This eliminates the race condition without a lock.

Reply:
Yeah composition > inheritance, if it's not worth to factor out common data (or even methods for that data), then that's fine.  If you say we cannot use mixed modes, do you you mean we cannot use PWM from PCA9685 for driving servo/esc and use PIGPIO, say for reading PWM simultaneously? Because this would be a drawback.

Reply:
I'm saying for the RPi.GPIO driver we can't mix BCM and BOARD pin numbers.  However, we can (and I have) mixed RPi.GPIO and PCA9685 pins.  You could also technically mix PIGPIO and PCA9685 pins.  Technically you could mix RPi.GPIO and PIGPIO, but that's not necessary and probably not advisable.

Reply:
Ok, cool, thanks for clarification.